fontdrasil/src/coords.rs

```diff
@@ -5,53 +5,143 @@
+    // <https://doc.rust-lang.org/1.74.0/nomicon/phantom-data.html#table-of-phantomdata-patterns>
+    space: PhantomData<fn() -> Space>,
+}
+
+/// A coordinate in design space.
+pub type DesignCoord = Coord<DesignSpace>;
+/// A coordinate in user space
+pub type UserCoord = Coord<UserSpace>;
+/// A coordinate in normalized space
+pub type NormalizedCoord = Coord<NormalizedSpace>;
+
+impl<Space> Coord<Space> {
+    /// Create a new coordinate.
+    ///
+    /// Note that we do *not* impl From because we want conversion to be explicit.
+    pub fn new(value: impl Into<OrderedFloat<f32>>) -> Self {
+        Coord {
+            coord: value.into(),
+            space: PhantomData,
+        }
+    }
+
+    pub fn into_inner(self) -> OrderedFloat<f32> {
+        self.coord
+    }
+
+    pub fn to_f32(&self) -> f32 {
+        self.coord.into_inner()
+    }
+
+    /// Convert this coordinate into the target space.
+    pub fn convert<ToSpace>(&self, converter: &CoordConverter) -> Coord<ToSpace>
+    where
+        Space: ConvertSpace<ToSpace>,
+    {
+        Space::convert_coord(*self, converter)
+    }
+}
 
 /// A set of per-axis coordinates that define a specific location in a coordinate system.
 ///
-/// E.g. a user location is a `Location<UserCoord>`. Hashable so it can do things like be
+/// E.g. a user location is a `Location<UserSpace>`. Hashable so it can do things like be
 /// the key for a map of sources by location.
-#[derive(Serialize, Deserialize, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
-#[serde(from = "LocationSerdeRepr<T>", into = "LocationSerdeRepr<T>")]
-// we need this trait bound here for our serde impl to work
-pub struct Location<T: Clone>(pub(crate) BTreeMap<Tag, T>);
+#[derive(Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Location<Space>(BTreeMap<Tag, Coord<Space>>);
+
+/// A location in [`DesignSpace`].
+pub type DesignLocation = Location<DesignSpace>;
+/// A location in [`UserSpace`].
+pub type UserLocation = Location<UserSpace>;
+/// A location in [`NormalizedSpace`].
+pub type NormalizedLocation = Location<NormalizedSpace>;
+
+// a little helper to generate methods on coord/location for specific conversions
+macro_rules! convert_convenience_methods {
+    ($space:ident, $fn_name:ident) => {
+        impl<Space> Coord<Space>
+        where
+            Space: ConvertSpace<$space>,
+        {
+            pub fn $fn_name(&self, converter: &CoordConverter) -> Coord<$space> {
+                self.convert(converter)
+            }
+        }
 
-pub type DesignLocation = Location<DesignCoord>;
-pub type UserLocation = Location<UserCoord>;
-pub type NormalizedLocation = Location<NormalizedCoord>;
+        impl<Space> Location<Space>
+        where
+            Space: ConvertSpace<$space>,
+        {
+            pub fn $fn_name(&self, axes: &HashMap<Tag, &Axis>) -> Location<$space> {
+                self.convert(axes)
+            }
+        }
+    };
+}
```

Comment:
I wish there wasn't a macro but I don't have a better idea.

Reply:
The alternative is just to write essentially the same code three times. Is your concern based on the insanity of macros in C? Rust's macros are much more constrained, and it makes it easier to refactor things in the future, since we'd only need to edit in one place.